Enhance documentation on layouting GLSP diagrams

ndoschek opened the pull request:
- Add markdownlint and cspell as recommended extensions
  - Add settings for markdownlint and prettier support for markdown and json files
- Add documentation for UI extensions
  - Add separate page for UI extensions
  - Documentation and implementation showcase
- Add documentation for Layouters
- Add documentation for default views

Contributed on behalf of STMicroelectronics.

planger commented on content/documentation/clientLayouting/content.md:
```diff
@@ -43,43 +43,295 @@ new GNodeBuilder()
+Additional custom layouters can be contributed by creating a layouter class that extends the `AbstractLayouter` and optionally provide custom options that extend the `AbstractLayoutOptions`.
+
+In the following example we show a simple custom layouter that extends the `VBoxLayouterExt` and provides a custom layout option,
+which increases all children sizes by this value in both width and height, the default value is `50px`.
```
```suggestion
which increases the widths and heights of all children by a specified value, the default value is `50` pixel.
```

planger commented on content/documentation/clientLayouting/content.md:
```diff
@@ -43,43 +43,295 @@ new GNodeBuilder()
 
+</details>
+
+For independent implementations that extend the `AbstractLayout`, it will be bound as follows:
```
```suggestion
Custom layouter, which don't overwrite existing ones should extend the `AbstractLayout` and can be registered for a certain layouter key, e.g. `MyCustomLayouter.KIND`, that is used in the `layout` property of compartments as follows:
```

planger commented on content/documentation/clientLayouting/content.md:
```diff
@@ -43,43 +43,295 @@ new GNodeBuilder()
+
+</details>
+
+Then the custom layouter is ready to use:
```
```suggestion
With that, the custom layouter is ready to be used by compartments in the graphical model:
```

planger commented on content/documentation/rendering/content.md:
```diff
@@ -80,6 +84,220 @@ The viewer queries all registered views and creates a new virtual DOM which is t
+##### CircularNodeView
+
+A `CircularNodeView` creates a round shape with a radius computed from the shape's size (by default it computes the radius by the minimum of the shape's width or height and divides that by 2).
+The computation of the radius can be overriden and adapted to custom needs.
```
Do you mean with "overriden" that it can be overriden by specifying the value in the GModelElement? I think this should be specified more clearly.

ndoschek replied:
Adressed in PR #57

planger commented on content/documentation/rendering/content.md:
```diff
@@ -80,6 +84,220 @@ The viewer queries all registered views and creates a new virtual DOM which is t
+
+##### ExpandButtonView
+
+The `ExpandButtonView` renders a SVG element in the shape of a triangle that allows exapandable parent elements to trigger expansion, for example to display further element information.
```
```suggestion
The `ExpandButtonView` renders a SVG element in the shape of a triangle that allows expandable parent elements to trigger expansion, for example to display further element information.
```

planger commented on content/documentation/rendering/content.md:
```diff
@@ -80,6 +84,220 @@ The viewer queries all registered views and creates a new virtual DOM which is t
+
+##### PreRenderedView
+
+The `PreRenderedView` visualizes a previously rendered piece of svg code as a separate SVG element.
```
```suggestion
The `PreRenderedView` visualizes a previously rendered piece of svg code as a separate SVG element.
This enables putting SVG code directly in the graphical model, which may be useful for including complex images for certain use cases.
However, usually it is recommended to create a dedicated element type and register a dedicated view, which produces custom SVG, as this yields more flexibility to take bounds, etc., into account.
```

planger commented on content/documentation/rendering/content.md:
```diff
@@ -80,6 +84,220 @@ The viewer queries all registered views and creates a new virtual DOM which is t
+
+##### RectangularNodeView
+
+A `RectangularNodeView` creates a rectangular shape based shape's size.
```
```suggestion
A `RectangularNodeView` creates a rectangular shape based on the shape's size.
```

planger commented on content/documentation/rendering/content.md:
```diff
@@ -80,6 +84,220 @@ The viewer queries all registered views and creates a new virtual DOM which is t
+
+##### SGraphView
+
+The `SGraphView` renders the base svg canvas for an SModel and triggers the rendering of its children.
```
```suggestion
The `SGraphView` renders the base SVG canvas for an SModel and triggers the rendering of its children.
```

planger commented on content/documentation/rendering/content.md:
```diff
@@ -80,6 +84,220 @@ The viewer queries all registered views and creates a new virtual DOM which is t
+##### SRoutingHandleView
+
+A `SRoutingHandleView` renders a circle shaped element that servers as routing point for routable elements (e.g. Edges).
+Its position is computed by either a register `EdgeRouterRegistry` or passed routing arguments.
```
```suggestion
Its position is computed either by a registered `EdgeRouterRegistry` or the routing arguments of the element.
```

planger commented on content/documentation/rendering/content.md:
````diff
@@ -80,6 +84,220 @@ The viewer queries all registered views and creates a new virtual DOM which is t
+The computation of the radius can be overriden and adapted to custom needs.
+
+```ts
+configureModelElement(
````
I'm not sure how useful the registration code is. I think it'd be more useful (but also more work), to show an example of creating such a node in the GModelFactory and show a screenshot of the result.

ndoschek replied:
Adressed in PR https://github.com/eclipsesource/graphical-lsp-website/pull/57

planger commented on content/documentation/rendering/content.md:
```diff
@@ -80,6 +84,220 @@ The viewer queries all registered views and creates a new virtual DOM which is t
+
+##### GIssueMarkerView
+
+A `GIssueMarkerView` renders a issue marker on top of shapes that have a validation issue.
```
```suggestion
A `GIssueMarkerView` renders an issue marker on top of shapes. This is used to show validation results on elements (see Model Validation).
```

planger commented on content/documentation/rendering/content.md:
```diff
@@ -80,6 +84,220 @@ The viewer queries all registered views and creates a new virtual DOM which is t
+
+##### RoundedCornerNodeView
+
+A `RectangularNodeView` creates a rectangular shape based shape's size and computes and renders the corners in a rounded way, based on the given options (i.e. `getClipPathInsets`).
```
I think the corner radius defaults to 0, so it is worth mentioning that the corner radius needs to be specified.
Also I'd mention all arguments and how they need to be passed.

This is related to the comment above that an example of creating a GModelElement that uses this type might be more helpful than showing the registration.

Also a link to the source code might be very helpful in general for all the views mentioned here.

ndoschek replied:
Thanks, yes I totally agree, I addressed this in PR https://github.com/eclipsesource/graphical-lsp-website/pull/57

planger commented on content/documentation/rendering/content.md:
````diff
@@ -80,6 +84,220 @@ The viewer queries all registered views and creates a new virtual DOM which is t
+```
+
+##### StructureCompartmentView
+
````
What is that?

ndoschek replied:
Oh sorry, I oversaw that section, I addressed this in PR https://github.com/eclipsesource/graphical-lsp-website/pull/57

planger commented on content/documentation/rendering/content.md:
```diff
@@ -80,6 +84,220 @@ The viewer queries all registered views and creates a new virtual DOM which is t
+
+##### ForeignObjectView
+
+The `ForeignObjectView` renders a box , which is usually contained by a default node view, which takes care of resizing and moving of the element.
```
```suggestion
The `ForeignObjectView` renders elements that are foreign to SVG, such as HTML, MathML, etc. as specified in their `namespace` and `code` property. Usually such an element is contained by a node view that enables features, such as resizing and moving of the element.
```

planger commented on content/documentation/rendering/content.md:
```diff
@@ -80,6 +84,220 @@ The viewer queries all registered views and creates a new virtual DOM which is t
+##### ForeignObjectView
+
+The `ForeignObjectView` renders a box , which is usually contained by a default node view, which takes care of resizing and moving of the element.
+The view then renders the provided code snippet, e.g. a `<pre>` element.
```
```suggestion
```

planger commented on content/documentation/rendering/content.md:
````diff
@@ -80,6 +84,220 @@ The viewer queries all registered views and creates a new virtual DOM which is t
+);
+```
+
+An example use case for using a `ForeignObjectView` is for example a multiline-text-box.
````
```suggestion
A common example use case for using a `ForeignObjectView` is to benefit from word wrapping support of HTML to show multiline text box.
```

planger commented on content/documentation/ui-extensions/content.md:
```diff
@@ -0,0 +1,174 @@
+  sticky = true
++++
+
+An UI Extension is a concept provided by the underlying framework Sprotty, which allows to display additional UI HTML elements on topf of a diagram.
```
```suggestion
A UI Extension is a concept provided by the underlying framework Sprotty, which allows to display additional UI HTML elements on top of a diagram.
```

planger commented on content/documentation/ui-extensions/content.md:
```diff
@@ -0,0 +1,174 @@
+
+An UI Extension is a concept provided by the underlying framework Sprotty, which allows to display additional UI HTML elements on topf of a diagram.
+
+It is often used to add UI controls to the diagram, good base examples are:
```
```suggestion
It is often used to add UI controls to the diagram. It is also frequently used in the core GLSP functionality:
```

planger commented on content/documentation/ui-extensions/content.md:
```diff
@@ -0,0 +1,174 @@
+To implement a custom UI extension, it is necessary to extend the base [`AbstractUIExtension`](https://github.com/eclipse/sprotty/blob/master/packages/sprotty/src/base/ui-extensions/ui-extension.ts).
+This super class provides a base HTML element (`containerElement`) which is then the base for the custom UI elements.
+It also provides utility methods for showing/hiding or handling focus of the extension.
+To add custom UI elements, the abstract methods `initializeContents` shall be implemented.
```
```suggestion
To add the actual UI elements to the DOM, the abstract method `initializeContents` must be implemented.
```

planger commented on content/documentation/ui-extensions/content.md:
```diff
@@ -0,0 +1,174 @@
+
+The [UIExtensionRegistry][uiex-registry] allows to register multiple UIExtensions per diagram.
+
+To enable an UI Extension the corresponding Action or Command shall be used, namely [`SetUIExtensionVisibilityAction`][uiex-registry] or [`SetUIExtensionVisibilityCommand`][uiex-registry].
```
```suggestion
To enable a UI Extension a [`SetUIExtensionVisibilityAction`][uiex-registry] or [`SetUIExtensionVisibilityCommand`][uiex-registry] with the respective UI Extension ID needs to be invoked.
```

planger commented on content/documentation/ui-extensions/content.md:
```diff
@@ -0,0 +1,174 @@
+To enable an UI Extension the corresponding Action or Command shall be used, namely [`SetUIExtensionVisibilityAction`][uiex-registry] or [`SetUIExtensionVisibilityCommand`][uiex-registry].
+
+As already mentioned, there is a broad use case for UI extensions, hence also their trigger events can of course vary.
+UI extensions can be enabled via context menu entries, mouse click events, keyboard events or changes of the diagram itself, for example the change of the diagram's `EditMode`.
```
```suggestion
UI extensions can be enabled practically with any event that can dispatch an action, e.g. via context menu entries, mouse click events, keyboard events or changes of the diagram itself, for example the change of the diagram's `EditMode`.
```

planger commented on content/documentation/ui-extensions/content.md:
````diff
@@ -0,0 +1,174 @@
+bind(TYPES.IUIExtension).toService(ButtonOverlay);
+```
+
+Activate ui extension, for example directly after all necessary initially Actions where dispatched in [GLSPDiagramWidget#`dispatchInitialActions()`](https://github.com/eclipse-glsp/glsp-theia-integration/blob/master/packages/theia-integration/src/browser/diagram/glsp-diagram-widget.ts).
````
```suggestion
To activate our UI extension, for example, alongside the initial actions, we can dispatch the `SetUIExtensionVisibilityAction` in [GLSPDiagramWidget#`dispatchInitialActions()`](https://github.com/eclipse-glsp/glsp-theia-integration/blob/master/packages/theia-integration/src/browser/diagram/glsp-diagram-widget.ts) after a certain delay.
```

planger reviewed (changes requested):
Thank you very much Nina! This adds urgently needed documentation and for sure will help newcomers to get started, but even more experienced developers to add more UI controls, etc.

Please find a few comments inline!

ndoschek commented:
Hi @planger, I uploaded a first update that addresses your review comments.
Regarding teh default view descriptions, I will add a second commit for that (and I also left those comments unresolved for now).

planger reviewed (commented):
Wow that's awesome! Thank you very much! Great work!

planger commented:
Do you want to merge the current version already and open a second PR or are you planning to append a commit to this PR?
I'm fine with either way!

ndoschek commented:
Thanks a lot @planger! I added the proper search site (when clicking on the search icon in the search fragment in the navigation bar) and fixed some minor typos and markdownlint issues.
Sounds good for merging that state, I will continue on integrating the remaining comments about the default views in a separate PR then. Thanks again!

planger reviewed (approved):
Thank you very much, Nina! Excellent work so far!
